### PR TITLE
Add configuration option for sync inference [3/3]

### LINKED ISSUE #1166: Sync map can be inferred from artifact Dockerfile
We already parse the source in a dockerfile's COPY and ADD, but we should also capture the destination.

```Dockerfile
FROM python:3.7-slim
CMD ["python", "-m", "flask", "run"]
ENV FLASK_DEBUG=1
ENV FLASK_APP=app.py

COPY requirements.txt .
RUN pip install -r requirements.txt
COPY *.py . # Should parse dst here
```
So that instead of writing this:
```yaml
  - image: gcr.io/k8s-skaffold/python-reload
    context: python
    sync:
      '*.py': .
```

Users could write this
```yaml
  - image: gcr.io/k8s-skaffold/python-reload
    context: python
    sync:
      '*.py':
```

However, there are many corner cases to watch out for. COPY can take multiple arguments to copy to the same destination folder, and there are rules of when the destination must be inferred as a directory or not.

Overlapping copies may be another corner case. Ultimately I think it will need to be modeled as a map of source files to possibly multiple locations.

Different artifact types can also provide different hints on how to fill in the sync map.

### PR DESCRIPTION
**Pull request by corneliusweig**
The sync feature is getting refurbished, see design proposal #1844 .

This PR implements step 3 of the implementation plan for the sync improvements. Notable changes include:

- add a new config option `sync.infer []string` to enable sync for some files
- add unit tests for creation of sync items
- add integration test for inferred sync mode

based-on #2084
Supersedes #1812
Fixes #1166, #1581
Related #1180, #1807
CC design shepherd @tejal29

### CODE REVIEW

**Comment by corneliusweig** [comment]
@tejal29 @balopat Just a polite reminder that sync inference is still not complete due to the missing `skaffold.yaml` adaptions from this PR.

**Comment by dgageot** [comment]
@corneliusweig What's the status on this PR? Is it good to be reviewed or are you still working on it? I'm happy to review

**Comment by corneliusweig** [comment]
@dgageot it is ready. Thanks for taking a look!

File: docs/content/en/docs/how-tos/filesync/_index.md [hunk 1]
```diff
@@ -15,6 +15,37 @@ The file copying is enabled by adding a `sync` section with _sync rules_ to the
 Under the hood, Skaffold creates a tar file with changed files that match the sync rules.
 This tar file is sent to and extracted on the corresponding containers. 
 
+### Inferred sync mode
+For docker artifacts, Skaffold knows to infer the desired destination from the artifact's `Dockerfile`.
```
Comment (dgageot) [inline docs/content/en/docs/how-tos/filesync/_index.md]: `knows` -> `knows how`?

**Comment by corneliusweig** [comment]
@tejal29 As you were the design shepherd: could you also review? Or find somebody to do that? Would be highly appreciated, as this PR has been ready for quite some time now.

File: pkg/skaffold/sync/sync.go [hunk 2]
```diff
@@ -86,11 +99,56 @@ func NewItem(a *latest.Artifact, e filemon.Events, builds []build.Artifact, inse
...
+	return &Item{Image: tag, Copy: toCopy, Delete: toDelete}, nil
+}
+
+func inferredSyncItem(a *latest.Artifact, e filemon.Events, builds []build.Artifact, provider DestinationProvider) (*Item, error) {
+	// deleted files are no longer contained in the syncMap, so we need to rebuild
+	if len(e.Deleted) > 0 {
+		return nil, nil
+	}
+
+	tag := latestTag(a.ImageName, builds)
```
Comment (nkubala) [inline pkg/skaffold/sync/sync.go]: nit: combine with the next line

```golang
if tag: latestTag(a.ImageName, builds); tag == "" {
  ...
}
```

File: pkg/skaffold/sync/sync.go [hunk 3]
```diff
@@ -86,11 +99,56 @@ func NewItem(a *latest.Artifact, e filemon.Events, builds []build.Artifact, inse
...
+	if tag == "" {
+		return nil, fmt.Errorf("could not find latest tag for image %s in builds: %v", a.ImageName, builds)
+	}
+
+	syncMap, err := provider()
+	if err != nil {
+		return nil, errors.Wrapf(err, "infer syncmap for image %s", a.ImageName)
```
Comment (nkubala) [inline pkg/skaffold/sync/sync.go]: nit: s/infer/inferring

File: pkg/skaffold/sync/sync.go [hunk 4]
```diff
@@ -86,11 +99,56 @@ func NewItem(a *latest.Artifact, e filemon.Events, builds []build.Artifact, inse
...
+	}
+
+	toCopy := make(map[string][]string)
+	for _, f := range append(e.Modified, e.Added...) {
+		relPath, err := filepath.Rel(a.Workspace, f)
+		if err != nil {
+			return nil, errors.Wrapf(err, "changed file %s can't be found relative to context %s", f, a.Workspace)
```
Comment (nkubala) [inline pkg/skaffold/sync/sync.go]: nit: `errors.Wrapf(err, "finding changed file %s relative to context %s", f, a.Workspace)`

File: pkg/skaffold/sync/sync.go (hunk 2 above)
Comment (corneliusweig) [reply pkg/skaffold/sync/sync.go]: It's also used in the return value in line 151, so that's not possible :(

**Comment by corneliusweig** [comment]
@nkubala addressed your comments and rebased on master (there were conflicts)

**Comment by nkubala** [comment]
@corneliusweig looks good except I think there's an issue with the config file getting passed to `TestDevSyncAPITrigger`, it's looking for `skaffold.yaml` but I think you changed it to `skaffold-infer.yaml`

**Comment by corneliusweig** [comment]
@nkubala you are right. That config needed to be adapted. Now let's wait for travis.

File: pkg/skaffold/sync/sync.go (hunk 2 above)
Comment (nkubala) [reply pkg/skaffold/sync/sync.go]: ah, didn't catch that.

**Comment by corneliusweig** [comment]
@nkubala CI is still failing. I now recall to also have refactored this test, because it had a loophole. I need to do the necessary adjustments tomorrow.

**Comment by corneliusweig** [comment]
@nkubala Ok, CI is finally green again. There were some hickups.

**Comment by nkubala** [comment]
@corneliusweig awesome, thank you for pushing this through!

**Pull request merged**